### PR TITLE
Add LABEL annotation and label node to BOX annotation

### LINKED ISSUE #26: Text annotations
Good stuff, you sure saved me a couple hours! Any updates on Text annotations?

Additionally, wouldn't it be neat to attach annotations to datasets? That way, annotations can be hidden when its related dataset is toggled.

Has this project been abandoned, or are there any plans for these updates or similar?

Thanks :)

### LINKED ISSUE #42: Box annotation labels
AFAICT, labels are not supported on box annotations, it would be nice if they were added.

### PR DESCRIPTION
**Pull request by stockiNail**
Fixes #26
Fixes #42

### Box annotation

Creates new `label` element to enable the label in the box  annotation.

### Label annotation
 
Creates new `LabelAnnotation` element to enable to render a text on the chart at a coordinate, calculated by the chart data values.

### Line annotation

Changes the line annotation in order to use the common methods for labelling and box drawing

### Ellipse annotation

Removes the inheritance from BOX annotation to maintain the same model for all annotations.

### Others

Moves common methods to **helpers** to reduce duplicated code:

1. getRectCenterPoint, to calculated the center point of a rectangle
1. getChartPoint, to get the coordinates of a point in the chartArea by xValue and yValue annotation options
1. getChartRect, to get the rectangle from the  chartArea by xMin, xMax, yMin and yMax annotation options

### TODO

- [x] types definition
- [X] documentation

### CODE REVIEW

**Review by kurkle — COMMENTED** [review]
Couple of comments, looking quite solid!

**Comment by kurkle** [comment]
| Filename | Size | Change |  |
| :--- | :---: | :---: | :---: |
| `dist/chartjs-plugin-annotation.esm.js` | 8.82 kB | +1.09 kB (+14%) | ⚠️ |
| `dist/chartjs-plugin-annotation.js` | 8.94 kB | +1.09 kB (+14%) | ⚠️ |
| `dist/chartjs-plugin-annotation.min.js` | 6.02 kB | +594 B (+11%) | ⚠️ |

**Comment by kurkle** [comment]
Off topic again, but should replace xPadding and yPadding with a `padding`

**Comment by stockiNail** [comment]
There are still other  points to fix in my TODO, in next days, but it’s important to me to know if this approach is better.

**Comment by stockiNail** [comment]
> Filename	Size	Change	
> `dist/chartjs-plugin-annotation.esm.js`	8.82 kB	+1.09 kB (+14%)	⚠️
> `dist/chartjs-plugin-annotation.js`	8.94 kB	+1.09 kB (+14%)	⚠️
> `dist/chartjs-plugin-annotation.min.js`	6.02 kB	+594 B (+11%)	⚠️

Could be an issue?

**Comment by kurkle** [comment]
> There are still other points to fix in my TODO, in next days, but it’s important to me to know if this approach is better.

I like this approach.

>>Filename Size Change
>>dist/chartjs-plugin-annotation.esm.js 8.82 kB +1.09 kB (+14%) ⚠️
>>dist/chartjs-plugin-annotation.js 8.94 kB +1.09 kB (+14%) ⚠️
>>dist/chartjs-plugin-annotation.min.js 6.02 kB +594 B (+11%) ⚠️

> Could be an issue?

No, its ok.

File: src/types/label.js
Comment (stockiNail) [reply src/types/label.js:34]: > I think I'd call it `this.labelRect` for consistency

@kurkle I purposely didn't use `labelRect` because the same attribute is used in line annotation but with different "meaning" (it's not ONLY the rect of the text but rect + padding and borderwidth). Here it is only the rect of the text.

Of course I can change but we have to pay attention that `label.labelRect` and `line.labelRect `contain different things

File: src/annotation.js
Comment (stockiNail) [reply src/annotation.js]: done

File: src/helpers.js
Comment (stockiNail) [reply src/helpers.js]: done

**Comment by kurkle** [comment]
Just a note, please don't change the x/yPadding in this PR, because that will be a breaking chage.

**Comment by stockiNail** [comment]
> Just a note, please don't change the x/yPadding in this PR, because that will be a breaking chage.

Yes, that si a breaking change

**Comment by stockiNail** [comment]
@kurkle a note from my side.

In the other PR, we had doubts about the use of `position` as option name. I propose `verticalAlign` (short `vAlign`).

Let me know because I don't have much fantasy for names..

**Comment by kurkle** [comment]
@etimberg thoughts about position / verticalAlign / vAlign?

**Comment by etimberg** [comment]
`verticalAlign` / `vAlign` are fine if these will always be in a vertical direction. I'd prefer `alignment` or `position` if we're making a generic interface that could apply in different directions

**Comment by stockiNail** [comment]
> `verticalAlign` / `vAlign` are fine if these will always be in a vertical direction. I'd prefer `alignment` or `position` if we're making a generic interface that could apply in different directions

@etimberg it's difficult to normalize "position" value, because in line annotation is used to position the label, with value start, center and end. Align is not present because not needed to align the label on the line.

In a box annotation, the label must be positioned horizontally (left, center, right) and vertically (top, center, bottom). The same for label annotation where the label is positioned based on the selected point.

I don't know if makes sense to have `position` property with different meaning (as is this PR implementation) among annotation types.

If all agree, I would propose for implementation:

| options | Line | Box | Label |
| :- | :- | :- | :- 
| position | start - center - end | - | - 
| align | - | left - center - right | left - center - right |
| vAlign | - | top - center - bottom | top - center - bottom |

Let me know if makes sense for you.

**Comment by kurkle** [comment]
🤔 align could be start/center/end because that works better with RTL. Which is likely requested at some point.

**Comment by etimberg** [comment]
That's a good point @kurkle on RTL support.

@stockiNail I can agree that proposal to treat lines differently than the others

**Comment by stockiNail** [comment]
> 🤔 align could be start/center/end because that works better with RTL. Which is likely requested at some point.

For RTL you should use `textAlign` and not `align`. `align` option positions ONLY the box where the label is placed and not the text of the label.

**Comment by kurkle** [comment]
What if the label on line is rotated, in that case align could be useful? baseLine could be another option usable with line annotations

**Comment by stockiNail** [comment]
@kurkle `rotation` is not implemented yet, in box and label annotations.  I have already evaluated but it could be more complex and I'd suggest to implement in another PR, if you agree.

About `baseline`,  would you change `align` to `baseline` or `position` to `baseline` ?

**Comment by stockiNail** [comment]
>

**Comment by stockiNail** [comment]
@kurkle @etimberg toady I thought about the names of the options.

I start thinking about line, where `position` is already implemented. At the end of the day, position can be used (with similar meaning) also in the other annotations. The position is related to the annotation start and end points.
For Label and Box annotation label, the vertical alignment is anyway needed, and I think what @kurkle suggest could make sense, using `baseline` as option name.

I have drawn some schemas to explain what we could implement:

## Position

### Line annoation label

LineLabelPosition

### Box annotation label

BoxLabelPosition

### Label annotation

LabelPosition

As you can see, in the label annotation, `end` means the label ends on selected point, `start` viceversa. 

## Baseline

### Box annotation label

BoxLabelBaseline

### Label annotation

LabelBaseline

What do you think?

The images, if makes sense to you, could be added to the documentation to explain the options usage (maybe also drawio files, if you like).

**Comment by kurkle** [comment]
Thanks @stockiNail for the images, it helps a lot.

From that I think the `position` as implemented, is the label position relative to the annotation element. 

It almost feels like we are re-inventing the wheel vs https://chartjs-plugin-datalabels.netlify.app/guide/positioning.html

I don't actually like baseline here, because its so tightly associated to font (https://developer.mozilla.org/en-US/docs/Glossary/baseline)

Considering we already have the `position`, we could just extend that to `position: {x: 'start', y: 'end'}`?

**Comment by stockiNail** [comment]
> Considering we already have the position, we could just extend that to position: {x: 'start', y: 'end'}?

For me, makes **a lot of sense**. I didn't think we could use an object to maintain the same `position` option, among all annotation elements.

As soon as I can, I'm going to start implementing it, hopefully to complete by tomorrow EOD.

**Comment by stockiNail** [comment]
@kurkle I removed `align` option and enabled `position` options as string or object, `{x: string, y: string}`.

Let me know if there is anything to fix. From my standpoint, there isn't anything else to add to this PR.

File: test/fixtures/label/scriptableOptions.js
Comment (kurkle) [inline test/fixtures/label/scriptableOptions.js:50]:
```suggestion
              content: (ctx, opts) => 'xValue: ' + opts.xValue + ', position: ' + JSON.stringify(opts.position),
```

File: test/fixtures/box/missingScale.js
Comment (kurkle) [inline test/fixtures/box/missingScale.js:23]:
```suggestion
              xMin: 'February',
```

**Review by kurkle — COMMENTED** [review]
Found some minor issues.

**Comment by etimberg** [comment]
Support an object sounds like a great solution to me!

File: test/fixtures/label/contentMultiline.js
Comment (stockiNail) [reply test/fixtures/label/contentMultiline.js:29]: I forgot to switch the control. Do you think we should include a picture in the documentation? I have the feeling some users will apply `start` thinking to have the label on the left

Comment (kurkle) [reply test/fixtures/label/contentMultiline.js:29]: Picture would not hurt, but that can be added later.

File: test/fixtures/label/positionVertical.js
Comment (stockiNail) [reply test/fixtures/label/positionVertical.js:25]: done

File: docs/.vuepress/config.js
Comment (stockiNail) [reply docs/.vuepress/config.js]: done

File: test/fixtures/label/contentMultiline.js
Comment (stockiNail) [reply test/fixtures/label/contentMultiline.js]: done

File: test/fixtures/label/positionHorizontal.js
Comment (stockiNail) [reply test/fixtures/label/positionHorizontal.js]: done

File: src/types/label.js
Comment (stockiNail) [reply src/types/label.js]: done

File: src/helpers.js
Comment (kurkle) [inline src/helpers.js:20]: Why this? Types still allow left/right too?

Comment (stockiNail) [reply src/helpers.js]: reapplied. I have removed because I thought only `start` and `end` were accepted.

File: types/label.d.ts
Comment (kurkle) [inline types/label.d.ts:119]: This was why I asked about the previous question. Need to have consistency. It's probably better to support also left/right too, for better UX (or is it DX in this case) .

**Review by kurkle — DISMISSED** [review]

File: types/label.d.ts
Comment (stockiNail) [reply types/label.d.ts:119]: fixed

**Comment by stockiNail** [comment]
@kurkle CodeClimate issue. I'm having a look how to solve "similar code"... Make sense?

**Comment by kurkle** [comment]
> @kurkle CodeClimate issue. I'm having a look how to solve "similar code"... Make sense?

Cc is a bit too sensitive with those, in my opinion. Too much generalization is not always good. So I'm fine with that kind of duplication.

**Comment by stockiNail** [comment]
any constraint for final approval and merge? Anything is missing?

**Comment by kurkle** [comment]
> any constraint for final approval and merge? Anything is missing?

waiting for @etimberg's review

**Comment by stockiNail** [comment]
> > any constraint for final approval and merge? Anything is missing?
> 
> waiting for @etimberg's review

Thank you @kurkle

File: test/fixtures/box/label.js
Comment (etimberg) [inline test/fixtures/box/label.js:68]: Is this supposed to be a function option?

**Review by etimberg — DISMISSED** [review]

File: test/fixtures/box/label.js
Comment (stockiNail) [reply test/fixtures/box/label.js:68]: the content is wrong... let me change the label and check the other test cases

**Review by etimberg — APPROVED** [review]

File: test/fixtures/box/label.js
Comment (stockiNail) [reply test/fixtures/box/label.js]: done! Thank you very much @etimberg

**Comment by stockiNail** [comment]
Milestone 1.2.0?

**Comment by stockiNail** [comment]
FYI and if you agree, after merge in the master, I'm going to rebase the other PR for polygon annotation and start working on padding as object (another new PR).
I think there will be to add test cases on all annotation types (for instance checks events) in order to increase the coverage (always if you agree).

**Pull request merged**
